mdata/notifierKafka/notifierKafka.go

```diff
@@ -66,26 +66,98 @@ func New(instance string, handler mdata.NotifierHandler) *NotifierKafka {
 }
 
 func (c *NotifierKafka) start() {
+	// offset is defined here because it is not used anywhere else
+	type offset struct {
+		// the actual kafka offset to use
+		offsetStart int64
+		// the time we use to attempt to get an offset (i.e. sarama.OffsetOldest)
+		offsetTime int64
+		// any errors returned from attempting to get an offset
+		offsetError error
+	}
+
 	pre := time.Now()
 	processBacklog := new(sync.WaitGroup)
+
+	// | scenario                   | offsetOldest  | offsetNewest | offsetTime
+	// ------------------------------------------------------------------------------------------------------------------------------
+	// | new empty partition        | error         | 0            | error
+	// | new with messages          | 0             | validOffset  | validOffset or error if offsetTime is earlier than first message
+	// | existing with messages     | validOffset   | validOffset  | validOffset
+	// | existing with no messages  | error         | validOffset  | error
+	// ------------------------------------------------------------------------------------------------------------------------------
+
 	for _, partition := range partitions {
-		var offsetTime int64
+		var startOffset int64
+		var validOffset bool
+		var offsetFromDuration int64
+
+		// make sure we are using a custom time offset, otherwise leave
+		// offsetFromDuration at 0
+		if offsetStr != "oldest" && offsetStr != "newest" {
+			offsetFromDuration = time.Now().Add(-1*offsetDuration).UnixNano() / int64(time.Millisecond)
+		}
+
+		offsets := map[string]*offset{
+			"oldest": {offsetTime: sarama.OffsetOldest},
+			"newest": {offsetTime: sarama.OffsetNewest},
+			"custom": {offsetTime: offsetFromDuration},
+		}
+
+		// get all of the offsets
+		for _, name := range []string{"newest", "custom", "oldest"} {
```

Comment:
I assume in most cases there won't be an error when querying the offset from Kafka. So is it necessary to always query it 3 times per partition, even when there's no error? Wouldn't it be more efficient to only query the other possible offsets when the initial one returned with an error?

Reply:
I did think about that, but in this case I don't think it matters. In the grand scheme of things when we talk about consuming the partitions a few milliseconds or seconds won't make a difference AFAIK. I'm up for it though.